### PR TITLE
[sn-platform][sn-platform-slim]: upgrade pulsar to 3.0 and enable 3.0 new features 

### LINKED ISSUE #1068: upgrade pulsar image to 3.0

### PR DESCRIPTION
**Pull request by ericsyh**
*(If this PR fixes a github issue, please add `Fixes #<xyz>`.)*

Fixes #1068 

*(or if this PR is one task of a github issue, please add `Master Issue: #<xyz>` to link to the master issue.)*

Master Issue: #<xyz>

### Motivation

*Explain here the context, and why you're making that change. What is the problem you're trying to solve.*

### Modifications

*Describe the modifications you've done.*

### Verifying this change

- [x] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Documentation

Check the box below.

Need to update docs? 

- [ ] `doc-required` 
  
  (If you need help on updating docs, create a doc issue)
  
- [x] `no-need-doc` 
  
  (Please explain why)
  
- [ ] `doc` 
  
  (If this PR contains doc changes)

### CODE REVIEW

File: charts/sn-platform-slim/values.yaml [hunk 1]
```diff
@@ -911,7 +911,9 @@ bookkeeper:
     extraOptions: []
     gcLoggingOptions: []
   custom: {}
-  configData: {}
+  configData: 
+    # Enable DirectIO by default
+    PULSAR_PREFIX_dbStorage_directIOEntryLogger: "true"
```
Comment (yaalsn) [inline charts/sn-platform-slim/values.yaml:916]: Pulsar 3.0 may lose data if it enables direct IO.  This PR will fix it. 

https://github.com/apache/bookkeeper/pull/4041

https://streamnative.slack.com/archives/C047JFTB9M3/p1689914711139779?thread_ts=1686858789.677249&cid=C047JFTB9M3

**Review by urfreespace — APPROVED** [review]

File: charts/sn-platform-slim/values.yaml [hunk 2]
```diff
@@ -1168,16 +1170,19 @@ broker:
...
     managedLedgerDefaultEnsembleSize: "3"
     managedLedgerDefaultWriteQuorum: "3"
     managedLedgerDefaultAckQuorum: "2"
     PULSAR_PREFIX_systemTopicEnabled: "true"
     PULSAR_PREFIX_topicLevelPoliciesEnabled: "true"
+    # Enable the new delayed message
+    delayedDeliveryTrackerFactoryClassName: "org.apache.pulsar.broker.delayed.BucketDelayedDeliveryTrackerFactory"
+    # Enable transaction buffer segmented snapshot
+    transactionBufferSegmentedSnapshotEnabled: "true"
+    transactionBufferSnapshotSegmentSize: "262144"
```
Comment (codelipenghui) [inline charts/sn-platform-slim/values.yaml:1182]: We don't need set this one, just use the default value that pulsar provided.

File: charts/sn-platform/values.yaml [hunk 3]
```diff
@@ -1243,16 +1245,19 @@ broker:
...
     managedLedgerDefaultEnsembleSize: "3"
     managedLedgerDefaultWriteQuorum: "3"
     managedLedgerDefaultAckQuorum: "2"
     PULSAR_PREFIX_systemTopicEnabled: "true"
     PULSAR_PREFIX_topicLevelPoliciesEnabled: "true"
+    # Enable the new delayed message
+    delayedDeliveryTrackerFactoryClassName: "org.apache.pulsar.broker.delayed.BucketDelayedDeliveryTrackerFactory"
+    # Enable transaction buffer segmented snapshot
+    transactionBufferSegmentedSnapshotEnabled: "true"
+    transactionBufferSnapshotSegmentSize: "262144"
```
Comment (codelipenghui) [inline charts/sn-platform/values.yaml:1257]: We don't need set this one, just use the default value that pulsar provided.

File: charts/sn-platform-slim/values.yaml [hunk 4]
```diff
@@ -1168,16 +1170,19 @@ broker:
...
-    # 2. You can use a k8s secret to reference AWS credentials. See `offload.s3.secret`
-    # AWS_ACCESS_KEY_ID: "[YOUR AWS ACCESS KEY ID]"
-    # AWS_SECRET_ACCESS_KEY: "[YOUR SECRET]"
     managedLedgerDefaultEnsembleSize: "3"
     managedLedgerDefaultWriteQuorum: "3"
     managedLedgerDefaultAckQuorum: "2"
     PULSAR_PREFIX_systemTopicEnabled: "true"
     PULSAR_PREFIX_topicLevelPoliciesEnabled: "true"
+    # Enable the new delayed message
+    delayedDeliveryTrackerFactoryClassName: "org.apache.pulsar.broker.delayed.BucketDelayedDeliveryTrackerFactory"
```
Comment (mattisonchao) [inline charts/sn-platform-slim/values.yaml:1179]: Well, I think maybe we should support dynamic calculate of the segment size or bucket size by cluster resources in the future. :) 

/cc @coderzc WDYT?

File: charts/sn-platform/values.yaml [hunk 5]
```diff
@@ -1243,16 +1245,19 @@ broker:
...
+    # Enable the new Load Balancer
+    loadManagerClassName: "org.apache.pulsar.broker.loadbalance.extensions.ExtensibleLoadManagerImpl"
```
Comment (Demogorgon314) [inline charts/sn-platform/values.yaml:1259]: To use `ExtensibleLoadManagerImpl`, we recommend upgrading the pulsar to 3.0.x before enabling the new load balancer.

https://pulsar.apache.org/docs/next/concepts-broker-load-balancing-migration/#migrate-from-modular-to-extensible-broker-load-balancer

cc @heesung-sn

File: charts/sn-platform-slim/values.yaml (hunk 4 above)
Comment (coderzc) [reply charts/sn-platform-slim/values.yaml:1179]: Yes, but we need to design a set of calculation formulas.

File: charts/sn-platform/values.yaml [hunk 6]
```diff
@@ -987,6 +987,8 @@ bookkeeper:
     gcLoggingOptions: []
   custom: {}
   configData: {}
+    # For Pulsar 3.0, please do not enable Direct IO. Needs a fix https://github.com/apache/bookkeeper/pull/4041
+    # PULSAR_PREFIX_dbStorage_directIOEntryLogger: "true"
```
Comment (hangc0276) [inline charts/sn-platform/values.yaml:991]: Please change the default value to `false`

Comment (ericsyh) [reply charts/sn-platform/values.yaml]: @hangc0276 fixed in https://github.com/streamnative/charts/pull/1083/commits/462b4a7cbddc93ff662fec9262f71d8d85237333

**Review by Demogorgon314 — APPROVED** [review]

**Review by tuteng — APPROVED** [review]

**Pull request merged**